### PR TITLE
feat: add authorization header

### PR DESCRIPTION
**Pull request by chkpwd**
Added authorization header to the authentication provider in efforts to auth using Basic Auth or Bearer Token via terraform provider.

### CODE REVIEW

**Review by Fuochi — COMMENTED** [review]
Hi @chkpwd, thanks for your first contribution. Just a couple of notes:
- Please fix the lint following CI suggestion
- It would be nice to test it once. Right now, all the tests run using api_key, which also validate the configuration, maybe you can add a step on `TestAccHostResource()` which uses this? but it's optional, I'll approve either way once the CI is working

File: internal/provider/provider.go
Comment (chkpwd) [reply internal/provider/provider.go]: Either one has to be populated. I don't think you can communicate to prowlarr without authorization?

Comment (uhthomas) [reply internal/provider/provider.go]: I may be misunderstanding, but I believe it currently says if either `key` or `authorization` is empty then it will fail. It should be valid for at least one of them to be empty, but not both.

Comment (chkpwd) [reply internal/provider/provider.go]: Oh, you're right. Fixing now.

File: internal/provider/provider.go
Comment (uhthomas) [inline internal/provider/provider.go:52]: The `but not both` is not enforced anywhere.

Comment (uhthomas) [inline internal/provider/provider.go:52]:
```suggestion
				MarkdownDescription: "Token for token-based authentication with Prowlarr. This is an alternative to using an API key. Set this via the `PROWLARR_AUTHORIZATION` environment variable. One of `authorization` or `api_key` must be provided, but not both.",
```

just a nit, and up to preference

File: internal/provider/provider.go
Comment (uhthomas) [inline internal/provider/provider.go:129]: Maybe remove the leading `.`? I don't know how errors render from this project, but the message before didn't have this.

File: internal/provider/provider.go
Comment (chkpwd) [reply internal/provider/provider.go:52]:
```go
	if key == "" && authorization == "" {
		resp.Diagnostics.AddError(
			"Missing Authentication Credentials",
			"Both 'api_key' and 'authorization' are empty. You must provide either an API key or an authorization token for Prowlarr authentication.",
		)
		return
	} else if key != "" && authorization != "" {
		resp.Diagnostics.AddError(
			"Conflicting Authentication Credentials",
			"Both 'api_key' and 'authorization' are provided. You must only provide one of these for Prowlarr authentication.",
		)
		return
	}
```
Looks good?

Comment (chkpwd) [reply internal/provider/provider.go:52]: fixed.

Comment (uhthomas) [reply internal/provider/provider.go:52]: Think so, but you can drop the `else if` part. A second, independent `if` statement is sufficient.

File: internal/provider/provider.go
Comment (chkpwd) [reply internal/provider/provider.go:129]:
```suggestion
			"Both 'api_key' and 'authorization' are empty. You must provide either an API key or an authorization token for Prowlarr authentication",
```
Remove it from the end of the sentence?

File: internal/provider/provider.go
Comment (chkpwd) [reply internal/provider/provider.go]: All done for now. I didn't add the test as I'm not too versed with the code. But once this is approved, I'll test it locally. If all goes well, I'll make the same changes to Sonarr and Radarr providers.

**Comment by Fuochi** [comment]
Hi @uhthomas, you left quite a review! thanks!
regarding this:
> A boolean field like UseAuthorizationHeader could instead use the existing code and infrastructure, without the complication

Honestly, I prefer @chkpwd approach of having 2 different fields, each one of them with its own meaning.

However, I agree with you that we could rethink the way it works a bit. From a user perspective, I don't think it should matter if I provide both headers as long as there is a clear order of preference. Which, in my opinion, could be:
`api_key` field > `authorization` field > `PROWLARR_API_KEY` env var > `PROWLARR_AUTHORIZATION` env var

So, maybe, we can avoid throwing the error if they're both present and instead set up only the one that is going to be used following that order. What do you think @chkpwd, @uhthomas?

**Comment by Fuochi** [comment]
@chkpwd I'll be without internet for one week, I'll check once I'm back. sorry.

**Review by Fuochi — CHANGES_REQUESTED** [review]
Hi @chkpwd,
can you implement this logic? https://github.com/devopsarr/terraform-provider-prowlarr/pull/196#issuecomment-1897841324
If not, no issue, I'll have a look at that later.
Either way, please, rebase and squash all your commits into one, then I'll merge!
Thanks.

**Review by Fuochi — APPROVED** [review]

**Pull request merged**
